dnscache.go

```diff
@@ -51,14 +51,22 @@ func (r *Resolver) Refresh(clearUnused bool) {
 	r.once.Do(r.init)
 	r.mu.RLock()
 	update := make([]string, 0, len(r.cache))
+	del := make([]string, 0, len(r.cache))
 	for key, entry := range r.cache {
 		if entry.used {
 			update = append(update, key)
 		} else if clearUnused {
-			delete(r.cache, key)
+			del = append(del, key)
 		}
 	}
 	r.mu.RUnlock()
+
+	for _, key := range del {
+		r.mu.Lock()
```

Comment:
You should move this lock/unlock outside of the loop.

Reply:
Also, add a condition that test for del len so you don't lock if there is nothing to delete.

Reply:
Done.  Thanks for the suggestion.